hilti/toolchain/src/compiler/global-optimizer.cc

```diff
@@ -0,0 +1,244 @@
+            dispatch(i);
+    }
+
+    void prune(Node& node) {
+        _stage = Stage::PRUNE;
+
+        bool modified = false;
+
+        for ( auto&& [hook_id, uses] : *_hooks ) {
+            // Linker joins are implemented via functions, so if we remove all
+            // functions data dependencies (e.g., needed for subunits) might
+            // get broken. Leave at least one function in unit so it gets emitted.
+            //
+            // TODO(bbannier): Explicitly express data dependencies in joins,
+            // see https://github.com/zeek/spicy/issues/918.
+            if ( std::get<2>(hook_id) == std::string("__str__") ) {
+                uses.defined = true;
+                continue;
+            }
+        }
+
+        for ( auto i : this->walk(&node) ) {
+            if ( auto x = dispatch(i) )
+                modified = modified && x;
+        }
+
+        if ( ! modified )
+            return;
+
+        prune(node);
+    }
+
+    result_t operator()(const type::struct_::Field& x, position_t p) {
+        if ( auto type_ = x.type().tryAs<type::Function>(); ! type_ )
+            return false;
+
+        auto field_id = x.id();
+
+        auto struct_type = typeID(p.parent().as<type::Struct>());
+        if ( ! struct_type )
+            return false;
+        auto&& [module_id, struct_id] = *struct_type;
+
+        auto type_ = p.findParent<declaration::Type>();
+        bool is_cxx = type_ && AttributeSet::find(type_->get().attributes(), "&cxxname");
+
+        auto ids = std::tuple<ModuleID, StructID, FieldID>{module_id, struct_id, field_id};
+
+        switch ( _stage ) {
+            case Stage::COLLECT: {
+                auto fn = x.childsOfType<Function>();
+                assert(fn.size() <= 1);
+
+                bool is_keep = ! fn.empty() && AttributeSet::find(fn.front().attributes(), "&keep");
+
+                // Record a declaration for this member.
+                (*_hooks)[ids].declared = true;
+
+                // If the member declaration is marked `&keep` mark it as implemented.
+                if ( is_keep )
+                    (*_hooks)[ids].defined = true;
+
+                // If the member declaration includes a body mark it as implemented.
+                if ( ! fn.empty() && fn.front().body() )
+                    (*_hooks)[ids].defined = true;
+
+                // If the unit is wrapped in a type with a `&cxxname`
+                // attribute its members are defined in C++ as well.
+                if ( is_cxx )
+                    (*_hooks)[ids].defined = true;
+
+                break;
+            }
+
+            case Stage::PRUNE: {
+                // Remove hooks without implementation.
+                if ( auto hook = _hooks->at(ids); ! hook.defined ) {
+                    HILTI_DEBUG(logging::debug::GlobalOptimizer,
+                                util::fmt("removing field for unused default hook %s::%s::%s", module_id, struct_id,
```

Comment:
```suggestion
                                util::fmt("removing field for unused hook %s::%s::%s", module_id, struct_id,
```

(not sure what `default` would mean)